components/discourse/common.js

```diff
@@ -0,0 +1,61 @@
+const crypto = require("crypto");
+const discourse = require("./discourse.app");
+const isEmpty = require("lodash.isempty");
+
+module.exports = {
+  props: {
+    http: "$.interface.http",
+    db: "$.service.db",
+    discourse,
+  },
+  dedupe: "unique",
+  hooks: {
+    async deactivate() {
+      const hookID = this.db.get("hookID");
+      await this.discourse.deleteHook({ hookID });
+    },
+  },
+  methods: {
+    // Generic activate method used in component-specific hooks
+    async activate(webhookPayload) {
+      const secret = this.discourse.generateSecret();
+      this.db.set("secret", secret);
+      const { id } = await this.discourse.createHook({
+        endpoint: this.http.endpoint,
+        secret,
+        ...webhookPayload,
+      });
+      this.db.set("hookID", id);
+    },
+    verifySignature(header, body) {
+      const algo = "sha256";
+      const hmac = crypto.createHmac(algo, this.db.get("secret"));
+      hmac.update(JSON.stringify(body), "utf-8");
+      if (header !== `${algo}=${hmac.digest("hex")}`) {
+        throw new Error(
+          "Discourse signature does not match computed signature"
+        );
+      }
+    },
+    generateMeta() {
+      throw new Error("Generate meta is not implemented");
+    },
+    validateEventAndEmit(event, eventName, key) {
+      const { body, headers } = event;
+      if (isEmpty(headers) || !headers["x-discourse-event-signature"]) {
+        console.log("Discourse signature header not present. Exiting");
```

Comment:
Maybe this should throw? Seems on the same level as the sig check failing vs. the event name filtering.